### DISCUSSION TITLE
basic-a2dp-audiokit lots of question marks

### DISCUSSION THREAD

**Discussion by John2009R** [pschatzmann/arduino-audio-tools#464] [Q&A] [answered]
Hi,
I have had success with some of the examples with my ESP32 Audio Kit V2.2 A210 (seems like a rare version).
I can run streams-audiokit-multioutput. with the deafault bord 5 so I think this proves the I2S config is right but not sure about I2C.
When I try basic-a2dp-audiokit it prints out the text below and then just prints question marks. No baud rate I try makes it into anything ineligible. I don't see a bluetooth device show up.
Any suggestions?
Thanks,
John

rst:0x1 (POWERON_RESET),boot:0x1f (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13864
load:0x40080400,len:3608
entry 0x400805f0
[I] AudioTypes.h : 54 - sample_rate: 44100
[I] AudioTypes.h : 55 - channels: 2
[I] AudioTypes.h : 56 - bits_per_sample: 16
[I] I2SConfig.h : 98 - rx/tx mode: TX_MODE
[I] I2SConfig.h : 99 - port_no: 0
[I] I2SConfig.h : 100 - is_master: Master
[I] I2SConfig.h : 101 - sample rate: 44100
[I] I2SConfig.h : 102 - bits per sample: 16
[I] I2SConfig.h : 103 - number of channels: 2
[I] I2SConfig.h : 104 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 106 - auto_clear:0
[I] I2SConfig.h : 113 - buffer_count:10
[I] I2SConfig.h : 114 - buffer_size:512
[I] I2SConfig.h : 118 - pin_bck: 14
[I] I2SConfig.h : 119 - pin_ws: 15
[I] I2SConfig.h : 120 - pin_data: -1515870811
[W] AudioKit.h : 54 - sd_active = false
[I] AudioKit.h : 202 - Channels 2->2
[I] AudioKit.h : 593 - void audio_tools::AudioKitStream::setupActions()
[W] AudioKit.h : 596 - Deactivating SPI because SD is not active
[I] AudioKit.h : 343 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 13 / logic: 0
[I] AudioActions.h : 52 - pin 13 -> INPUT_PULLUP
[I] AudioActions.h : 42 - ActionLogic::add pin: 39 / logic: 2
[I] AudioActions.h : 55 - pin 39 -> INPUT
[I] AudioKit.h : 343 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 18 / logic: 3
[I] AudioActions.h : 55 - pin 18 -> INPUT
[I] AudioKit.h : 343 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 5 / logic: 3
[I] AudioActions.h : 55 - pin 5 -> INPUT
???????????????????????

**Comment by pschatzmann** [comment]
try to comment out the   kit.processActions();
If this helps you have selected the wrong board and have confilics with the buttons in the current setting

**Reply by John2009R** [reply]
Thanks for the quick reply Phil (and for sharing all these projects).
I tried your suggestion and some of the output changed:
[I] AudioKit.h : 343 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 13 / logic: 0
[I] AudioActions.h : 52 - pin 13 -> INPUT_PULLUP
[I] AudioActions.h : 42 - ActionLogic::add pin: 39 / logic: 2
[I] AudioActions.h : 55 - pin 39 -> INPUT
[I] AudioKit.h : 343 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 18 / logic: 0
[I] AudioActions.h : 52 - pin 18 -> INPUT_PULLUP
[I] AudioKit.h : 343 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 5 / logic: 0
[I] AudioActions.h : 52 - pin 5 -> INPUT_PULLUP

...but still the same end result with the backwards question marks. I left the dip switches as it was when it arrived 11100. 
Any other ideas?
Thanks

**Comment by pschatzmann** [comment]
I am using 01100 as indicated in https://github.com/pschatzmann/arduino-audiokit/wiki

**Comment by John2009R** [comment]
Oh, I didn't look at that SD page since I haven't try to use it yet, but setting to 01100 didn't change anything for this sketch.

My board isn't one of the listed supported ones and searching for others with A210 I only found one dead-end hit.
Is it reasonable to assume my I2S config is correct given the streams-audiokit-multioutput sketch works?
What about I2C? 
Thanks again

**Comment by pschatzmann** [comment]
if you get sound then the I2S and I2C settings are correct. 
It is strange that you dont get any proper error message. Did you Activate the ESP32 logging in the Arduino Menu ?
I am using ES32 Dev Module as board in Arduino

**Reply by John2009R** [reply]
I am also using ESP32 Dev Board and just went with the defaults.  I never noticed that debug option.
When I set it to Info the program became too big.  I tried changing the partition scheme to 8MB with spiffs. That goes into reset loop.
Then I set it to Huge App as I saw referenced elsewhere here and then I got more logs and... it worked! I mean I was able to connect to it on my phone and send music to it. 
I don't know why that would have fixed it. I am suspecting my board might be flakey. Sometimes it doesn't work and I reset it and then it works.

**Comment by pschatzmann** [comment, marked as answer]
I am using Huge App most of the time. A2DP is getting quite big and when we load audio data into PROGMEM this is the only way to go...
You should get an error however if the sketch does not fit!

**Reply by John2009R** [reply]
I did some experimenting and found my board would go off into space (printing gibberish) on the basic-a2dp-audiokit on about 90% of resets. The problem seemed to be in various parts of i2c configuring but usually as the Bluetooth was being setup.

I tried changing the i2c frequency from 100KHz to 10KHz and then it ran properly for about 5 resets in a row. I thought my issue was solved and I posted about it. You may have gotten an email. Then I found it didn't really change anything and that sometimes it works a few times in a row and sometimes it fails repeatedly so I deleted my post.

When it works it seems stable, I listened to about 6 songs from my phone without issue. I also uncommented kit.processActions and the volume control buttons work on the speaker. 

I did a screen capture incase my description is not clear: https://youtu.be/gLDV06RtJP0

Have you heard of others having unreliable initialization like this?  I am ready to chalk it up to a bad board and if you agree I will consider ordering another one.
Thanks

**Comment by pschatzmann** [comment]
This is the first time I hear about this issue, so it seems that so far you are the only one.
It is difficult to say if this is related to your individual board or your version.

By default the system uses the Arduino i2c wire implementation. You can try if it is more stable if you switch to the ESP32 I2C API.
This setting can be found in AudioKitSettings.h

It would also be interesting to hear if this is only related to A2DP or if you see this issue for other sketches as well...

**Reply by John2009R** [reply]
I tried disabling the wire implementation of i2c and the sketch worked through 10 resets and I thought it was solved (again). Then I came back to it later and when I powered it up it did not work and I had to reset it about 10 times to get it to work.

I tried the guitar effects sketch and it worked (very cool) but it was not stable. It was working for 30 seconds and then would hang and just make a hum and print out garbage. After resetting it would work again for a while. It seems the problems come outside of doing I2C configuration and it is probably just my board.

Anyways, it has been fun to play with. If I think of something I want to make more permanent I will order another one and hope for better luck.
Thanks
